perception/navigator_vision/nodes/pinger_finder.py

```diff
@@ -0,0 +1,169 @@
+#!/usr/bin/env python
+import rospy
+import numpy as np
+from numpy import linalg as npl
+import tf
+from navigator_msgs.srv import FindPinger, SetFrequency, SetFrequencyResponse
+from hydrophones.msg import ProcessedPing
+from geometry_msgs.msg import Point, PoseStamped, Pose
+from visualization_msgs.msg import Marker
+from std_srvs.srv import SetBool, SetBoolResponse
+import navigator_tools
+
+target_freq = 35000
+freq_tol = 1000
+min_amp = 200
+max_samples = 200 # will delete entries with least amplitude after we get pass this count
+line_array = np.empty((0, 4), float)
+sample_amplitudes = np.empty((0, 0), float)
+pinger_position = Point(0, 0, 0)
+viz_pub = rospy.Publisher("/visualization_marker", Marker, queue_size=10)
+point_distance = 10  # p1 is aling the direction of the pinger but we do not know the actual
+                     # range p1 will be adjusted to be 'point_distance' away from p0
+debug_arrow_id = 0
+listen = False
+
+# line_arr represents the equation of a 2d line. Structure: [x1, y1, x2, y2]
+# Calculates the point in the plane with the least cummulative distance to
+# every line in line_arr
+def LS_intersection(line_arr):
+    global pinger_position
+    def line_norm(line_end_pts):
+        assert len(line_end_pts) == 4
+        begin = line_end_pts[:2]
+        end = line_end_pts[2:]
+        diff = end - begin
+        return npl.norm(diff)
+
+    num_lines = line_arr.shape[0]
+    begin_pts = line_arr[:, :2]
+    end_pts = line_arr[:, 2:4]
+    diffs = end_pts - begin_pts
+    norms = np.apply_along_axis(line_norm, 1, line_arr)
+    norms = norms.reshape(norms.shape[0], 1)
+    perp_unit_vecs = np.apply_along_axis(lambda x: np.array([[0, -1], [1, 0]]).dot(x), 1, diffs / norms)
+    A_sum = np.zeros((2, 2))
+    Ap_sum = np.zeros((2, 1))
+
+    for x, y in zip(begin_pts, perp_unit_vecs):
+        begin = x.reshape(2, 1)
+        perp_vec = y.reshape(2, 1)
+        A = perp_vec.dot(perp_vec.T)
+        Ap = A.dot(begin)
+        A_sum += A
+        Ap_sum += Ap
+
+    res = npl.inv(A_sum).dot(Ap_sum)
+    pinger_position = Point(res[0], res[1], 0)
+
+
+def find_pinger(arg):
+    if line_array.shape[0] < 2:
+        return None
+    LS_intersection(line_array)
+    return {'pinger_position' : pinger_position, 'num_samples' : line_array.shape[0]}
+
+
+# receives pings and adds the representation of a line between the hydrophone array and the pinger
+# to our line_arr linear equation system
+def ping_cb(processed_ping):
+    global line_array, sample_amplitudes
+    if not listen:
+      return
+    print "Got processed ping message:\n{}".format(processed_ping)
+    if abs(processed_ping.freq - target_freq) < freq_tol and processed_ping.amplitude > min_amp:
+        print "\x1b[34mTrustworthy pinger heading\x1b[0m"
+        trans, rot = None, None
+        try:
```

Comment:
Probably should only enclose the TF stuff in the `try`. Also you may want to print out some kind of error notification so that the error message was caused by an error.